core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/orm.go

```diff
@@ -0,0 +1,78 @@
+package upkeepstate
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/lib/pq"
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type orm struct {
+	chainID *utils.Big
+	q       pg.Q
+}
+
+type persistedStateRecord struct {
+	UpkeepID            *utils.Big
+	WorkID              common.Hash
+	CompletionState     uint8
+	BlockNumber         int64
+	IneligibilityReason uint8
+	InsertedAt          time.Time
+}
+
+// NewORM creates an ORM scoped to chainID.
+func NewORM(chainID *big.Int, db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *orm {
+	return &orm{
+		chainID: utils.NewBig(chainID),
+		q:       pg.NewQ(db, lggr.Named("ORM"), cfg),
+	}
+}
+
+// InsertUpkeepState is idempotent and sets upkeep state values in db
+func (o *orm) InsertUpkeepState(state persistedStateRecord, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+
+	query := `INSERT INTO evm_upkeep_states (evm_chain_id, work_id, completion_state, block_number, inserted_at, upkeep_id, ineligibility_reason)
+	  VALUES ($1::NUMERIC, $2::BYTEA, $3, $4, $5, $6::NUMERIC, $7)
+	    ON CONFLICT (evm_chain_id, work_id)
+	    DO NOTHING`
+
+	return q.ExecQ(query, o.chainID, state.WorkID, state.CompletionState, state.BlockNumber, state.InsertedAt, state.UpkeepID, state.IneligibilityReason)
+}
+
+// SelectStatesByWorkIDs searches the data store for stored states for the
+// provided work ids and configured chain id
+func (o *orm) SelectStatesByWorkIDs(workIDs []common.Hash, qopts ...pg.QOpt) (states []persistedStateRecord, err error) {
+	q := o.q.WithOpts(qopts...)
+
+	ids := make([][]byte, 0)
+
+	for _, id := range workIDs {
+		ids = append(ids, id.Bytes())
+	}
+
+	err = q.Select(&states, `SELECT upkeep_id, work_id, completion_state, block_number, ineligibility_reason, inserted_at
+	  FROM evm_upkeep_states
+	  WHERE work_id = ANY(SELECT unnest($1::BYTEA[]) addr) AND evm_chain_id = $2::NUMERIC`, pq.Array(ids), o.chainID)
```

Comment:
This should not be necessary, you should be able to do:
```suggestion
	  WHERE work_id = ANY($1) AND evm_chain_id = $2::NUMERIC`, pq.Array(ids), o.chainID)
```

Reply:
I'll try it again. It didn't provide any results from the select statement.